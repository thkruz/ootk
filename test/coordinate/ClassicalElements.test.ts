import { ClassicalElements, EpochUTC, J2000, Kilometers, KilometersPerSecond, Radians, Vector3D }
  from './../../src/main';

describe('ClassicalElements', () => {
  const epoch = EpochUTC.fromDateTime(new Date('2024-01-14T14:39:39.914Z'));
  let elements: ClassicalElements;

  beforeEach(() => {
    elements = new ClassicalElements({
      epoch,
      semimajorAxis: 6943.547853722985 as Kilometers,
      eccentricity: 0.0011235968124658146,
      inclination: 0.7509087232045765 as Radians,
      rightAscension: 0.028239555738616327 as Radians,
      argPerigee: 2.5386411901807353 as Radians,
      trueAnomaly: 0.5931399364974058 as Radians,
    });
  });

  // can be constructed with valid parameters
  it('should construct a ClassicalElements object with valid parameters', () => {
    expect(elements).toMatchSnapshot();
  });

  // can convert to EquinoctialElements
  it('should convert ClassicalElements to EquinoctialElements', () => {
    const equinoctialElements = elements.toEquinoctialElements();

    expect(equinoctialElements).toMatchSnapshot();
  });

  // can propagate to a new epoch
  it('should propagate ClassicalElements to a new epoch', () => {
    const propEpoch = EpochUTC.fromDateTime(new Date('2024-01-15T14:39:39.914Z'));
    const propagatedElements = elements.propagate(propEpoch);

    expect(propagatedElements).toMatchSnapshot();
  });

  // can calculate mean motion
  it('should calculate the mean motion of ClassicalElements', () => {
    const meanMotion = elements.meanMotion;

    expect(meanMotion).toMatchSnapshot();
  });

  // can calculate period
  it('should calculate the period of ClassicalElements', () => {
    const period = elements.period;

    expect(period).toMatchSnapshot();
  });

  // can calculate apogee
  it('should calculate the apogee of ClassicalElements', () => {
    const apogee = elements.apogee;

    expect(apogee).toMatchSnapshot();
  });

  // can calculate perigee
  it('should calculate the perigee of ClassicalElements', () => {
    const perigee = elements.perigee;

    expect(perigee).toMatchSnapshot();
  });

  // can calculate orbit regime
  it('should calculate the orbit regime of ClassicalElements', () => {
    const orbitRegime = elements.getOrbitRegime();

    expect(orbitRegime).toMatchSnapshot();
  });

  // can convert to PositionVelocity
  it('should convert ClassicalElements to PositionVelocity', () => {
    const positionVelocity = elements.toPositionVelocity();

    expect(positionVelocity).toMatchSnapshot();
  });

  // can convert toString
  it('should convert ClassicalElements to string', () => {
    const string = elements.toString();

    expect(string).toMatchSnapshot();
  });

  // can calculate inclination in degrees
  it('should calculate inclination in degrees', () => {
    const inclination = elements.inclinationDegrees;

    expect(inclination).toMatchSnapshot();
  });
  // can calculate right ascension in degrees
  it('should calculate right ascension in degrees', () => {
    const rightAscension = elements.rightAscensionDegrees;

    expect(rightAscension).toMatchSnapshot();
  });
  // can calculate argument of perigee in degrees
  it('should calculate argument of perigee in degrees', () => {
    const argPerigee = elements.argPerigeeDegrees;

    expect(argPerigee).toMatchSnapshot();
  });
  // can calculate true anomaly in degrees
  it('should calculate true anomaly in degrees', () => {
    const trueAnomaly = elements.trueAnomalyDegrees;

    expect(trueAnomaly).toMatchSnapshot();
  });

  // can create from StateVector
  it('should create ClassicalElements from StateVector', () => {
    const stateVector = new J2000(
      EpochUTC.fromDateTime(new Date(1705109326817)),
      new Vector3D(1538.223335842895 as Kilometers, 5102.261204021967 as Kilometers, 4432.634965003577 as Kilometers),
      new Vector3D(
        -7.341518909379302 as KilometersPerSecond,
        0.6516718453998644 as KilometersPerSecond,
        1.7933882499861993 as KilometersPerSecond,
      ),
    );

    const classicalElements = ClassicalElements.fromStateVector(stateVector);

    expect(classicalElements).toMatchSnapshot();
  });
});
