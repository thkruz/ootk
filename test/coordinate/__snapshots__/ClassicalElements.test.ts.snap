// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClassicalElements should calculate argument of perigee in degrees 1`] = `145.4534258954243`;

exports[`ClassicalElements should calculate inclination in degrees 1`] = `43.02390063917958`;

exports[`ClassicalElements should calculate right ascension in degrees 1`] = `1.6180073591471598`;

exports[`ClassicalElements should calculate the apogee of ClassicalElements 1`] = `580.3415297612491`;

exports[`ClassicalElements should calculate the mean motion of ClassicalElements 1`] = `0.0010911808567933578`;

exports[`ClassicalElements should calculate the orbit regime of ClassicalElements 1`] = `"Low Earth Orbit"`;

exports[`ClassicalElements should calculate the perigee of ClassicalElements 1`] = `564.7380332899556`;

exports[`ClassicalElements should calculate the period of ClassicalElements 1`] = `95.96920113444682`;

exports[`ClassicalElements should calculate true anomaly in degrees 1`] = `33.984415021959016`;

exports[`ClassicalElements should construct a ClassicalElements object with valid parameters 1`] = `
ClassicalElements {
  "argPerigee": 2.5386411901807353,
  "eccentricity": 0.0011235968124658146,
  "epoch": EpochUTC {
    "posix": 1705243179.914,
  },
  "inclination": 0.7509087232045765,
  "mu": 398600.4415,
  "rightAscension": 0.028239555738616327,
  "semimajorAxis": 6943.547853722985,
  "trueAnomaly": 0.5931399364974058,
}
`;

exports[`ClassicalElements should convert ClassicalElements to EquinoctialElements 1`] = `
EquinoctialElements {
  "I": 1,
  "a": 6943.547853722985,
  "epoch": EpochUTC {
    "posix": 1705243179.914,
  },
  "h": 0.0006107793657340771,
  "k": -0.0009430897960304828,
  "lambda": 3.15876545174161,
  "mu": 398600.4415,
  "p": 0.01112918196689249,
  "q": 0.3939942790541112,
}
`;

exports[`ClassicalElements should convert ClassicalElements to PositionVelocity 1`] = `
Object {
  "position": Vector3D {
    "x": -6935.3813042612455,
    "y": -146.1261328371035,
    "z": 46.43907991639276,
  },
  "velocity": Vector3D {
    "x": 0.07740377357485445,
    "y": -5.543955093986134,
    "z": -5.174123893746047,
  },
}
`;

exports[`ClassicalElements should convert ClassicalElements to string 1`] = `
"[ClassicalElements]
  Epoch: 2024-01-14T14:39:39.914Z
  Semimajor Axis (a):       6943.5479 km
  Eccentricity (e):         0.0011236
  Inclination (i):          43.0239°
  Right Ascension (Ω):      1.6180°
  Argument of Perigee (ω):  145.4534°
  True Anomaly (ν):         33.9844°"
`;

exports[`ClassicalElements should create ClassicalElements from StateVector 1`] = `
ClassicalElements {
  "argPerigee": 1.7497725502749213,
  "eccentricity": 0.0006908917666211547,
  "epoch": EpochUTC {
    "posix": 1705109326.817,
  },
  "inclination": 0.7503068041317154,
  "mu": 398600.4415,
  "rightAscension": 0.17555503341584255,
  "semimajorAxis": 6935.754028093152,
  "trueAnomaly": 5.749771010517163,
}
`;

exports[`ClassicalElements should propagate ClassicalElements to a new epoch 1`] = `
ClassicalElements {
  "argPerigee": 2.5386411901807353,
  "eccentricity": 0.0011235968124658146,
  "epoch": EpochUTC {
    "posix": 1705329579.914,
  },
  "inclination": 0.7509087232045765,
  "mu": 398600.4415,
  "rightAscension": 0.028239555738616327,
  "semimajorAxis": 6943.547853722985,
  "trueAnomaly": 0.6234422142843716,
}
`;
